### ISSUE TITLE
Feature request: cache based on content of files

### ISSUE THREAD

**Issue by 7rulnik** [sverweij/dependency-cruiser#715]
### Context

We already have [cache](https://github.com/sverweij/dependency-cruiser/blob/develop/doc/cli.md#--cache-use-a-cache-to-speed-up-cruising-experimental). It speeds up cruising massively. But it uses git to get modified files between runs. Usually, we have whole git tree, but sometimes we don't. For example, in CI people usually use git depth 1 for faster checkout ([gitlab for example](https://docs.gitlab.com/ee/ci/large_repositories/index.html#shallow-cloning)). So in this case we don't have git tree and can't use cache.

### Expected Behavior

It would be nice if we can use cache even if we don't have a whole git tree.

### Current Behavior

Currently, cache works only if we have SHA1 in git tree and if we don't we will run whole cruising from the start.

### Possible Solution

We can provide different strategies for caching or switch only for caching based on file content.
TypeScript uses file content for the [incremental feature](https://www.typescriptlang.org/tsconfig/incremental.html). Prettier, eslint, stylelint supports [--cache-strategy](https://eslint.org/docs/latest/user-guide/command-line-interface#--cache-strategy) flag. It allows choosing between strategies: modified time or hash of file content.

So on the first run, we can save the hash of file content into the cache.
And on the second we can go through modules, read them and compare actual hash of content with cached one.

So an implementation could be kinda like this:
```js
async function getChangedModules() {
    const changedModules = []
    for (const file of manifest.modules) {
        if (!file.couldNotResolve && !file.coreModule) {
            // We can use something like https://yarnpkg.com/package/p-limit here for concurrency
            const content = await fs.readFile(__dirname + '/' + file.source)
            const hashSum = crypto.createHash('sha256');
            hashSum.update(content);
            const hex = hashSum.digest('hex');
            if (hex !== file.hash) {
                changedModules.push(file.source)
            }
        }
    }

    return changedModules
}
```

On m1 pro 13 it takes 1s for 7200 modules.

So I think we can drop git for detecting changes and just use metadata and content like other tools or even always use content hashing stategy.

**Comment by sverweij** [comment]
Hi @7rulnik thanks for this suggestion! 

When implementing the cache feature my initial strategy was using hashes of file content. I decided against it because the strategy with git had both superior performance and a more straightforward implementation. As a 'backup' strategy (when `git clone` with a limited depth is called for or it isn't possible/ practical to use its `--shallow-since=` option) it could be a viable option - and indeed not _that_ hard to implement. I'm rummaging around in that area anyway for other reasons, so I'll see what I can whip up.

**Comment by sverweij** [comment]
Hi @7rulnik 

I've made a first (working) sketch of the functionality in the [feature/non-git-caching](https://github.com/sverweij/dependency-cruiser/tree/feature/non-git-caching) branch to understand the problem space and released it as `dependency-cruiser@12.5.0-beta-1` on npmjs (shasum: `718a91058569ef73956e8f467df529611dff9cb1`) so it's easier to install in other repo's (& on other machines) and that way work out the kinks. If you want to you can kick its tires (if you do - I'm _very_ interested in any feedback you might have).

Usage
- in .dependency-cruiser.js, in the `options` section somewhere add ...

```javascript
  // ...
  cache: {
     strategy: 'content'
  } 
  // ...
```

... and run dependency-cruiser as normal (recommended to not pass the `--cache` command line switch yet when this is engaged). 

> b.t.w. a --cache-strategy command line option might come, but I've concentrated on
> understanding the problem space first. 


h3. some background

> "We don't do these things because they're easy. 
> We do these things because we thought they were easy"
>
>    --- _free after John F. Kennedy_


It took a bit more effort than I had foreseen. Apart from some changes to make room for an additional cache strategy the algorithm needed some interesting additions.

To invalidate/ update a cache without using _git_ we have to cater for:
- _modules that are in the cache that have changed (or are deleted)_ - this is more or less what's covered in the algorithm above 
- _modules that are _not_ in the cache (_yet_ or _ever_)_ - for this we need to scan all files in the working directory and compare  them to the cache (with some optimizations e.g. applying .gitignore and any exclude or includeOnly filters that might be part of the current  dependency-cruiser options).


With the extra disk access (& computational intensity of determining checksums) it's slower and more memory intensive than the git-based 'metadata' strategy. For the cache 'freshness' step in dependency-cruiser's self scan (509 modules, 1295 dependencies) I have these averages:
- 262ms vs 72ms (3.6x) on a warm disk cache; 
- 400ms vs 72ms (5.5x) on a cold one; 
- ~13Mb heap usage vs ~2Mb for dependency-cruiser's self-scan. 

(pass `--progress performance-log` on the dependency-cruiser cli to see these numbers).

However, it's better than nothing in situations where the `metadata` strategy won't work. 

h3. next steps

- Try it out on some code bases (react, yarn, and some bigger private ones) to gain a bit of experience.
- Bend the sketched code into shape and/ or remove it and replace it with something properly engineered.
- Test-automate the hell out of it.
- Release in v12.5.0

**Comment by 7rulnik** [comment]
@sverweij I will take a look at this next week. Was busy with bag in Safari for whole week :D
BTW looks awesome, thanks!

**Comment by sverweij** [comment]
hi @7rulnik likely the content based strategy will land in release 12.5.0, which I hope to release tonight (GMT+1). In any case let me know what think (/ find)! 

FYI - a `--cache-strategy` command line option is coming up as well (see #724) which might make it in release 12.5.0 as well.

**Comment by sverweij** [comment]
FYI: I've just published `dependency-cruiser@12.5.0` with both features in ([12.5.0 release notes](https://github.com/sverweij/dependency-cruiser/releases/tag/v12.5.0)).

**Comment by 7rulnik** [comment]
@sverweij just tested it. Looks like I faced a bug or something.

After the first run I see a bunch of files with `"changeType": "added"`.

The second run takes almost the same time. And all non-js files marked as `deleted`:
```json
      {
        "name": "client/src/style.scss",
        "changeType": "deleted"
      }
```

The third run works as fast as expected, non-js files stay `deleted`

**Comment by 7rulnik** [comment]
Also performance is great. With this codebase:
```json
    "totalCruised": 7636,
    "totalDependenciesCruised": 24487,
```
it takes less than 2 seconds.

**Comment by sverweij** [comment]
> @sverweij just tested it. Looks like I faced a bug or something.
> 
> After the first run I see a bunch of files with `"changeType": "added"`.
> 
> The second run takes almost the same time. And all non-js files marked as `deleted`:
> 
> ```json
>       {
>         "name": "client/src/style.scss",
>         "changeType": "deleted"
>       }
> ```
> 
> The third run works as fast as expected, non-js files stay `deleted`

Hi @7rulnik thanks for this feedback! 

Indeed if you experienced this 'double run' than I consider that as a bug. I encountered this as well with `12.5.0-beta-1` but fixed it in one of the later beta's (+ in the regular 12.5.0 release); in later versions I couldn't reproduce  it  anymore. If you found  it in a later version I'm very interested to reproduce it.

=> what is  the version you encountered this in?

**Comment by 7rulnik** [comment]
@sverweij I faced it with 12.5.0

**Comment by sverweij** [comment]
Thanks!

I did find a lot of scars files on a react project with status _deleted_ as
well, so I can at least reproduce that one (I know where and how to fix it
- but as I'm currently AfK it'll have to wait a bit).

I'll see if I can reproduce the 'double run' on a clean version of that
same project - I might need to come back with you on that in case I can't
though.

**Comment by sverweij** [comment]
Hi @7rulnik - I was able to reproduce the double run  as well - it had  the same root cause as the superfluous _deleted_ change entries. Fix in PR #727 .

**Comment by sverweij** [comment]
I've published dependency-cruiser@12.5.1-beta-1 (with the fix from the PR) to npmjs.

**Comment by sverweij** [comment]
Now also published as the normal _latest_ release: dependency-cruiser@12.5.1 (thanks again for the validation note you left in the PR!)
